### ISSUE TITLE
"unregistered collection used in transaction" erroneously thrown when "read" collections are declared (with allowImplicit=false) -- forced to declare as "write" for now

### ISSUE THREAD

**Issue by hbCyber** [arangodb/arangodb#1967]
## Environment
- ArangoDB 3.0.2
- OS: Ubuntu 14.04.4 LTS 
- I am executing, via the Java driver, a Javascript transaction which contains AQL queries and anonymous graph functionality

## Steps to reproduce
*Please note that I have simplified my own code to produce a straightforward bug example. I understand that some of the logic is silly/useless in this transaction. Quality is not representative of my actual product.*

You can initialize the database collections with this snippet:
```
db._createDocumentCollection('account');
db._createDocumentCollection('moneyProfile');
db._createDocumentCollection('ledgerTransaction');
db._createDocumentCollection('moneyTransaction');

db._createEdgeCollection('ledgerTransactionAccount');
db._createEdgeCollection('moneyTransactionAccount');
```

Then, manually insert an *account* record and *moneyProfile* record with the same key, "test1".
```
db.account.insert({ '_key': 'test1' });
db.moneyProfile.insert({ '_key': 'test1' });
```

Now, for the upcoming JS transaction, you will include these flags:
```
allowImplicit: false

write: ledgerTransaction
write: ledgerTransactionAccount
write: moneyProfile
write: moneyTransaction
write: moneyTransactionAccount

read: account
```

Here is the _action_ Javascript transaction code that will cause the issue, if you execute it with the *params* object containing _'accountId': 'test1'_
```
function(params) {
    // Require(s)
    var db = require('org/arangodb').db;
	
    // Extract parameter(s)
    var accountId = params['accountId'];
	
    // Declare utility function(s)
    var generateSillyId = function() {
        return (Date.now() +
            Math.floor((Math.random() * 1000) + 1)).toString();
    };
	
    // Get account
    var account = db._query('FOR r IN account FILTER r._key == @accountId LIMIT 1 RETURN r', {
        'accountId': accountId
    }).toArray()[0];
	
    // Get money profile
    var moneyProfile = db._query('FOR r IN moneyProfile FILTER r._key == @accountId LIMIT 1 RETURN r', {
        'accountId': accountId
    }).toArray()[0];
	
    // Update money profile
    db._update('moneyProfile/' + accountId, {
        'test': true,
        'testDate': Date.now()
    });
    
	// Generate transaction ID
    var moneyTxId;
    do moneyTxId = generateSillyId();
    while (db._query('FOR r IN moneyTransaction FILTER r._key == @txId LIMIT 1 RETURN r', {
            'txId': moneyTxId
        }).toArray().length > 0);
    
	// Store money transaction
    db.moneyTransaction.insert({
        '_key': moneyTxId,
        'timestamp': Date.now(),
        'test': true
    });
    
	// Link to account
    db.moneyTransactionAccount.insert({
        '_from': 'moneyTransaction/' + moneyTxId,
        '_to': 'account/' + accountId
    });
    
	// Generate ledger transaction ID
    var ledgerTxId;
    do ledgerTxId = generateSillyId();
    while (db._query('FOR r IN ledgerTransaction FILTER r._key == @txId LIMIT 1 RETURN r', {
            'txId': ledgerTxId
        }).toArray().length > 0);
    
	// Store ledger transaction
    db.ledgerTransaction.insert({
        '_key': ledgerTxId,
        'timestamp': Date.now(),
        'test': true
    });
    
	// Link to account
    db.ledgerTransactionAccount.insert({
        '_from': 'ledgerTransaction/' + ledgerTxId,
        '_to': 'account/' + accountId
    });
    
	// Completion
    return {
        'success': true
    };
}
```

As you can see, that code never modifies the *account* collection itself. Still, unless I declare the *account* collection as 'write', the error is thrown:

_[1652] unregistered collection used in transaction_

**Comment by jsteemann** [comment]
The above transaction works with latest 3.0 (branch `3.0`, 9953557b5dee9bad50fe66453682834e41e442a). I will check with 3.0.3

**Comment by Simran-B** [comment]
I tried the following in arangosh, using the current devel branch:

```js
db._executeTransaction({
    collections: {
        allowImplicit: false,
        write: ["ledgerTransaction","ledgerTransactionAccount","moneyProfile","moneyTransaction","moneyTransactionAccount"],
        read: "account"
    },
    params: {accountId: "test1"},
    action: function (params) {
        // Require(s)
        var db = require('org/arangodb').db;

        // Extract parameter(s)
        var accountId = params['accountId'];

        // Declare utility function(s)
        var generateSillyId = function() {
            return (Date.now() +
                Math.floor((Math.random() * 1000) + 1)).toString();
        };

        // Get account
        var account = db._query('FOR r IN account FILTER r._key == @accountId LIMIT 1 RETURN r', {
            'accountId': accountId
        }).toArray()[0];

        // Get money profile
        var moneyProfile = db._query('FOR r IN moneyProfile FILTER r._key == @accountId LIMIT 1 RETURN r', {
            'accountId': accountId
        }).toArray()[0];

        // Update money profile
        db._update('moneyProfile/' + accountId, {
            'test': true,
            'testDate': Date.now()
        });

        // Generate transaction ID
        var moneyTxId;
        do moneyTxId = generateSillyId();
        while (db._query('FOR r IN moneyTransaction FILTER r._key == @txId LIMIT 1 RETURN r', {
                'txId': moneyTxId
            }).toArray().length > 0);

        // Store money transaction
        db.moneyTransaction.insert({
            '_key': moneyTxId,
            'timestamp': Date.now(),
            'test': true
        });

        // Link to account
        db.moneyTransactionAccount.insert({
            '_from': 'moneyTransaction/' + moneyTxId,
            '_to': 'account/' + accountId
        });

        // Generate ledger transaction ID
        var ledgerTxId;
        do ledgerTxId = generateSillyId();
        while (db._query('FOR r IN ledgerTransaction FILTER r._key == @txId LIMIT 1 RETURN r', {
                'txId': ledgerTxId
            }).toArray().length > 0);

        // Store ledger transaction
        db.ledgerTransaction.insert({
            '_key': ledgerTxId,
            'timestamp': Date.now(),
            'test': true
        });

        // Link to account
        db.ledgerTransactionAccount.insert({
            '_from': 'ledgerTransaction/' + ledgerTxId,
            '_to': 'account/' + accountId
        });

        // Completion
        return {
            'success': true
        };
    }
});
```

I had to add `params: {accountId: "test1"}` to avoid *Cannot read property 'accountId' of undefined*, but then it returns `{"success": true}`.

**Comment by hbCyber** [comment]
Thanks guys, so happy to hear this will be resolved in 3.0.4, that's amazing!

**Comment by jsteemann** [comment]
3.0.4 released today.
